### PR TITLE
[SPARK-43049][SQL] Use CLOB instead of VARCHAR(255) for StringType for Oracle JDBC

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Use CLOB instead of VARCHAR(255) for StringType for Oracle JDBC

### Why are the changes needed?

- Fix insufficient length issue when storing a spark string to oracle.
- Make room for Spark VarcharType mapping 

### Does this PR introduce _any_ user-facing change?

yes, Using APIs, such as DDL and `df.write.jdbc`, with oracle to store string will result in CLOB columns.
### How was this patch tested?

new tests.

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @dongjoon-hyun @cloud-fan thanks

**Review by cloud-fan — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks, merged to master

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/OracleIntegrationSuite.scala [hunk 1]
```diff
@@ -106,17 +106,26 @@ class OracleIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCTes
     var t = spark.table(tbl)
     var expectedSchema = new StructType().add("ID", DecimalType(10, 0), true, defaultMetadata)
     assert(t.schema === expectedSchema)
-    sql(s"ALTER TABLE $tbl ALTER COLUMN id TYPE STRING")
+    sql(s"ALTER TABLE $tbl ALTER COLUMN id TYPE LONG")
     t = spark.table(tbl)
-    expectedSchema = new StructType().add("ID", StringType, true, defaultMetadata)
+    expectedSchema = new StructType().add("ID", DecimalType(19, 0), true, defaultMetadata)
```
Comment (dongjoon-hyun) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/OracleIntegrationSuite.scala:111]: `LONG` is `DecimalType(19, 0)`?

**Review by dongjoon-hyun — COMMENTED** [review]
+1, late LGTM. Thank you!

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/OracleIntegrationSuite.scala (hunk 1 above)
Comment (yaooqinn) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/OracleIntegrationSuite.scala:111]: Yes, Oracle Built-in integrals are `NUMBER(p, 0)`, the `LONG` there is a deprecated alternative for character strings.

Comment (dongjoon-hyun) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/OracleIntegrationSuite.scala:111]: Got it. Thank you, @yaooqinn .

**Comment by barenko** [comment]
How can I force to use varchar2 instead clob?

I have a table with index in the varchar column (ROW_ID) and Oracle does not allow index on clobs.

**Comment by cloud-fan** [comment]
How about we revert it first? This brings perf issues for all use cases but only benefits corner cases where the string value is larger than 255.

We can have a config to control the varchar length for oracle dialect.

**Comment by yaooqinn** [comment]
> This brings perf issues for all use cases but only benefits corner cases where the string value is larger than 255.

Is this a result of inertia that prefers string to varchar in DDLs with Oracle jdbc data source?

**Comment by cloud-fan** [comment]
CLOB is slower than VARCHAR2 in Oracle, especially when the string value is small, which is a common case. @barenko  also raised a valid issue that CLOB has problems with index.

**Comment by yaooqinn** [comment]
We have the createTableColumnTypes option as a workaround. So, this is mainly about which type shall be the default?

**Comment by HyukjinKwon** [comment]
I think the problem is about the performance. Actually one guy from open source (that is not related to my employer) pinged me about this too

**Comment by yaooqinn** [comment]
> If the limit is exceeded, the excess data is silently discarded.

According to the JDBC standard, it can aslo be a correctness issue.

**Comment by cloud-fan** [comment]
@yaooqinn are you sure? IIUC we have CHAR/VARCHAR support for JDBC data source and Spark will perform length check during writing.

My point is this is kind of a breaking change for perf regression and index-related issues.

**Comment by yaooqinn** [comment]
@cloud-fan Yes, I have verified that we have a length check. It also seems not a big deal to create a string column but result in a varchar one. I'm +1 to revert this

**Comment by sadikovi** [comment]
I opened the revert PR https://github.com/apache/spark/pull/44452.

**Comment by gatorsmile** [comment]
@yaooqinn @cloud-fan @dongjoon-hyun  Any change like this must be documented in the migration guide.

**Comment by yaooqinn** [comment]
> Any change like this must be documented in the migration guide.

Thank you @gatorsmile.
It appears to be a performance regression. Do we need a migration guide for this instead of the PR that caused a behavior change with the write path length check?

**Pull request closed without merging**